content/post/practical_ml_2.md

```diff
@@ -0,0 +1,55 @@
+ 
+Consider our spam filter. Say that we have manually identified a number of spammy and non-spammy messages and proceed to build a model that, based on our historic data, does an excellent job of classifying spam. **Our model is only as good at identifying what spammy events look like, as the extent of diverse spammy events exist in our data set**. In other words, if we get a new example of spam that looks completely different from anything we have yet seen, then our model is less likely to catch it. 
+ 
+Spam filters are a special class of problems because spam is created by malcontents. These individuals (or groups) continuously try to poke holes in our system. They have a strong incentive to ensure that future spam looks different from past spam. What's more, these bad actors can detect what patterns our model has identified as likely to be spam, and adjust their methods accordingly. As a result, our future data will **always** look different from the past.
+ 
+Spam is an extreme example of a changing data distribution. However, the notion that future data will look different from your training data is a reality that pervades all types of problems. Factors that can impact your data distribution include:
+ 
+* *Seasonality*: Events around Christmas time look very different from events in the summer.
+* *Marketing efforts*: Different types of campaigns will attract different types of users, with different behaviors. 
+* *Changes in Regulatory landscape*: [GDPR](https://en.wikipedia.org/wiki/General_Data_Protection_Regulation?oldformat=true) is a prime example of this. What happens when we have to delete user data that our model has been leveraging?
+* *Product launches*: It is very common for tech startups to pivot unexpectedly. This wreaks havoc for our models.
+ 
+In order to combat the reality of a changing data distribution, the Data Scientist must retrain her model at regular intervals, perform feature engineering as needed and potentially change the entire algorithm that underlies her model. While we can strive to minimize our model's generalization error, we must recognize that our job is not complete just because we hit deploy.
+ 
+## Technical considerations
+ 
+Unless the Data Scientist's model is deployed on production, it usually runs on a Data Warehouse and likely does not interact with the application. This may result in the Data Scientist cutting some corners; it is not uncommon to sacrifice model optimization for increased accuracy. After all, if a model is not used in production then stakeholders don't need real-time results, making optimization a "nice to have". In addition, Data Scientists are often not trained in software engineering best practices. A Data Scientist’s model is usually a script having minimal interaction with a larger application. The range of a model’s inputs and outputs are much more narrow and easily controlled.
```

Comment:
"model optimization" could mean a lot of things here (one of which is "accuracy" 😄). Do you mean the performance characteristics, ie how quickly the model can make predictions?